src/eduid/webapp/idp/other_device.py

```diff
@@ -0,0 +1,229 @@
+            ttl=ttl,
+        )
+        res = self.save(state)
+        logger.debug(f'Save {state.state_id} result: {res}')
+        logger.info(f'Created other-device state: {state.state_id}')
+        logger.debug(f'   Full other-device state: {state.to_json()}')
+        return state
+
+    def abort(self, state: OtherDevice) -> Optional[OtherDevice]:
+        """
+        Abort a state.
+        """
+        if state.state not in [OtherDeviceState.NEW, OtherDeviceState.IN_PROGRESS]:
+            return None
+        _state_val = state.state.value
+        state.state = OtherDeviceState.ABORTED
+
+        result = self._coll.replace_one({'_id': state.obj_id, 'state': _state_val}, state.to_dict(), upsert=True)
+        logger.debug(
+            f'Aborted OtherDevice {state} in the db: '
+            f'matched={result.matched_count}, modified={result.modified_count}, upserted_id={result.upserted_id}'
+        )
+        if not result.acknowledged:
+            return None
+        return state
+
+    def grab(self, state: OtherDevice, device2_ref: str) -> Optional[OtherDevice]:
+        """
+        Grab a state, on device 2. This has to be an atomic operation to ensure two devices (one attacker and one
+        victim) can't have pending_requests pointing at this very same OtherDevice state. Otherwise, the attacker
+        can retrieve the response_code after the victim logs in.
+        """
+        if state.state != OtherDeviceState.NEW:
+            return None
+
+        if state.device2.ref is not None:
+            return None
+
+        _state_val = state.state.value
+        state.state = OtherDeviceState.IN_PROGRESS
+        state.device2.ref = device2_ref
+
+        result = self._coll.replace_one({'_id': state.obj_id, 'state': _state_val}, state.to_dict(), upsert=True)
+        logger.debug(
+            f'Grabbed OtherDevice {state} in the db: '
+            f'matched={result.matched_count}, modified={result.modified_count}, upserted_id={result.upserted_id}'
+        )
+        if not result.acknowledged:
+            return None
+        return state
+
+    def logged_in(self, state: OtherDevice, eppn: str, credentials_used: List[UsedCredential]) -> Optional[OtherDevice]:
+        """
+        Finish a state, on device 2.
+        """
+        if state.state != OtherDeviceState.IN_PROGRESS:
+            return None
+
+        if not state.device2.ref:
+            return None
+
+        if (state.eppn and state.eppn != eppn) or not eppn:
+            logger.error(f'Can\'t record use other device as finished for eppn {eppn} (state has eppn {state.eppn})')
+
+        _state_val = state.state.value
+        state.state = OtherDeviceState.LOGGED_IN
+        state.eppn = eppn
+        state.device2.credentials_used = credentials_used
+        state.device2.response_code = make_short_code()
+
+        result = self._coll.replace_one({'_id': state.obj_id, 'state': _state_val}, state.to_dict(), upsert=True)
```

Comment:
Again, shouldn't this fail if the state i missing/expired?

Reply:
fixed above